File: Week 9/X48980_en/permutations_1.cpp
/*
PERMUTATIONS(S)
    let T be an empty stack (of characters)
    for i = 1 to |S| do
        push s[i] onto T
    let L be an empty set (of strings)
    insert an empty string in L
    while T is not empty do
        pop from T the top element x
        let L' be an empty set (of strings)
        for all w in L do
            for i = 1 to |w| + 1 do
                insert w[1,...,i - 1] + x + w[i,...,|w|] in L'
        L = L'
    sort L
    return L
*/


#include <iostream>
#include <vector>
#include <set>
#include <stack>
using namespace std;


set<string> permutations_1(string s) {
    stack<string> t;
    for (int i = 0; i < s.length(); i++) {
        t.push(s.substr(i, 1));
    }
    set<string> l;
    l.insert("");
    while (!t.empty()) {
        string x = t.top();
        t.pop();
        set<string> l1;
        for (auto w : l) {
            for (int i = 0; i <= w.length(); i++) {
                l1.insert(w.substr(0, i) + x + w.substr(i, w.length() - i));
            }
        }
        l = l1;
    }
    return l;
}


int main() {
    string s;
    cin >> s;
    set<string> res = permutations_1(s);
    for (auto w : res) {
        cout << w << endl;
    }
    return 0;
}

File: Week 9/X48980_en/permutations_1.py
'''
PERMUTATIONS(S)
    let S be an empty stack (of characters)
    for i = 1 to |S| do
        push s[i] onto S
    let L be an empty set (of strings)
    insert an empty string in L
    while S is not empty do
        pop from S the top element x
        let L' be an empty set (of strings)
        for all w in L do
            for i = 1 to |w| + 1 do
                insert w[1,...,i - 1] + x + w[i,...,|w|] in L'
        L = L'
    sort L
    return L
'''


import sys


def permutations(s):
    t = list()
    for i in range(len(s)):
        t.append(s[i])
    l = set()
    l.add("")
    while len(t) > 0:
        x = t.pop()
        l1 = set()
        for w in l:
            for i in range(len(w) + 2):
                l1.add(w[:i] + x + w[i:])
        l = l1
    A = list()
    for w in l:
        A.append(w)
    A = sorted(A)
    return A


for w in permutations(sys.stdin.read()[:-1]):
    print(w)


    

